modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java

```diff
@@ -17,84 +17,104 @@
 import org.apache.ignite.internal.tostring.S;
-import org.apache.ignite.internal.util.IgniteUtils;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * Data storage manager.
  */
 public class DataStorageManager implements IgniteComponent {
-    /** Mapping: {@link StorageEngine#name} -> {@link StorageEngine}. */
+    /** Mapping: {@link StorageEngineFactory#name} -> {@link StorageEngine}. */
     private final Map<String, StorageEngine> engines;
 
     /**
      * Constructor.
      *
      * @param clusterConfigRegistry Register of the (distributed) cluster configuration.
      * @param storagePath Storage path.
-     * @throws StorageException If there are duplicates of the data storage engine.
+     * @param engineFactories Storage engine factories.
+     * @throws IllegalStateException If there are duplicates of the data storage engine.
      */
     public DataStorageManager(
             ConfigurationRegistry clusterConfigRegistry,
-            Path storagePath
+            Path storagePath,
+            Iterable<StorageEngineFactory> engineFactories
     ) {
-        this.engines = StreamSupport.stream(engineFactories().spliterator(), false)
-                .map(engineFactory -> engineFactory.createEngine(clusterConfigRegistry, storagePath))
+        engines = StreamSupport.stream(engineFactories.spliterator(), false)
                 .collect(toUnmodifiableMap(
-                        StorageEngine::name,
-                        identity(),
-                        (storageEngine1, storageEngine2) -> {
-                            throw new StorageException(String.format(
-                                    "Duplicate key [key=%s, engines=%s]",
-                                    storageEngine1.name(),
-                                    List.of(storageEngine1, storageEngine2)
-                            ));
-                        }
+                        StorageEngineFactory::name,
+                        engineFactory -> engineFactory.createEngine(clusterConfigRegistry, storagePath)
                 ));
     }
 
+    /**
+     * Constructor, overloads {@link DataStorageManager#DataStorageManager(ConfigurationRegistry, Path, Iterable)} with loading factories
+     * through a {@link ServiceLoader}.
+     */
+    public DataStorageManager(
+            ConfigurationRegistry clusterConfigRegistry,
+            Path storagePath
+    ) {
+        this(clusterConfigRegistry, storagePath, ServiceLoader.load(StorageEngineFactory.class));
+    }
+
     /** {@inheritDoc} */
     @Override
-    public void start() {
+    public void start() throws StorageException {
         engines.values().forEach(StorageEngine::start);
     }
 
     /** {@inheritDoc} */
     @Override
-    public void stop() throws Exception {
-        IgniteUtils.closeAll(engines.values().stream().map(engine -> engine::stop));
+    public void stop() throws StorageException {
+        engines.values().forEach(StorageEngine::stop);
```

Comment:
Ok, what was the problem with the old method? New one is worse, can you guess why?

Reply:
Fix it